### DISCUSSION TITLE
Gosh, this repo is really a mess

### DISCUSSION THREAD

**Discussion by ManiMozaffar** [ultrafunkamsterdam/undetected-chromedriver#1234] [Ideas]
Ok things that should've been fixed in this repo:

1. Using __del__, this repo has connection leak since when you started using __del__ without knowing what is deferred finalization and how messing up with __del__ can actually result in connection leak, memory leak or port exhaustion. You don't have to close chrome when script stop, nobody asked for this feature. people can literally call a method and quit or close the driver base on their needs.

2. Poorly implementation of downloading chromedriver

3. No github action and workflow to respect best CI practices. Perhaps you can run test to see if docker version of repo, or headless or headfull is detectable. You can even put trigger on chrome releases, and when a chrome is released then you can run this repo with latest chrome to see if it affect or not. if it does, creates an issue that chrome is detectable on that version

4. Why selenium? for serious, you could have used playwright, better support, better community, modern syntax, less overhead, more options.

5. Not adopted to latest selenium version, I've seen you've written some codes that was added as feature in selenium in later versions. 

6. You hate contributors, I've mentioned many times that if you do not have free time then that's what open source is for. but you don't comment on PRs, you don't even check them after months,  you just work on your local machine with your own code. Use others, for the repo's sake, and your sake.

**Comment by ManiMozaffar** [comment]
@ultrafunkamsterdam 
Hey Leon, if I make a PR fixing at least those python mistakes, alongside with some workflows, you'd merge it? so I wouldn't be wasting my time at least, like it happened last time :) Received no feedback from you!

Here's my suggestions:
1. Errors shouldn't be silent, I've dealt with connection leak because you've ignored errors almost everywhere in the code base.
2. There are some weird code I see, like you named a lambda function? I mean lambda is there to be nameless!
3. Weird comments all around the code, instead of explaining  it in docstring.
4. Weird namespaces, like "t" "d" and ... 
5. No test written for the repository! Test could be written to check if targeted wesbites are detecting current uc's implemention or not.
6. Fixing unused imports, and import structure which doesn't follow Pylint best practices.
7. Fixing type annotations all around the code. 
8. Providing nicer docs, at this point with 5k stars a documentation is something necessary.

**Reply by krjii** [reply]
do you have a pull request our for the connection leak?

**Comment by ManiMozaffar** [comment]
Yep found it what caused connection leak :) 
Does this now() - t ever gets bigger than 3? 
That's why CDP calls were never garbage collected.

```py
def __del__(self):
        if self._custom_exe_path:
            return
        else:
            timeout = 3
            t = time.monotonic()
            now = lambda: time.monotonic()
            while now() - t > timeout:
                try:
                    if self.user_multi_procs:
                        break
                    os.unlink(self.executable_path)
                    logger.debug("successfully unlinked %s" % self.executable_path)
                    break
                except (OSError, RuntimeError, PermissionError) as error:
                    logging.error(error)
                    time.sleep(0.01)
                    continue
                except FileNotFoundError as error:
                    logging.error(error)
                    break
```

**Comment by xAffan** [comment]
You should make undetected chromedriver for playwright

**Comment by jstoxrocky** [comment]
yes plz ^ @ManiMozaffar

**Reply by krjii** [reply]
would this be a point at where interested contributors fork?

**Comment by ManiMozaffar** [comment]
I'm not gonna create fork from this lib and maintain it myself, for 5 reasons:
- Contributors aren't welcome here, many PR aren't even reviewed.
- Code base is badly designed
- Codebase doesn't have test and CI/CD.
- Owner wrote some weird coding stuff in his coding, which result in port leakage and port exhaustion (see below ref), and doesn't seem to care.
- 700 Issues :) for a 6k stars lib that's too much.

https://github.com/ultrafunkamsterdam/undetected-chromedriver/blob/master/undetected_chromedriver/reactor.py
Reason of port leakage:
https://github.com/python/cpython/issues/104688

Some reasons you should avoid code base:
- Possible misses of request capture due to POORLY IMPLEMENTATION of async with hard coded sleeps 
- 0% test coverage
- Poor error handling
- Lack of type annotations
- Raise statements that would be never raised
- if you enable type annotations in your coding edittor, you would see he's catching an exception which is a subclass of above catched exceptions, so basically that would never be caught haha. and then there's a logic for that catch!! I think it was for the place where he wanted to remove a folder from os.
- I'm not even sure why he decided to use asyncio here out of context? wouldn't it be better to just use the normal threading only?

And I'm not even sure if it's thread safe.
I don't even like selenium at this point.
Best you can do guys is to just uninstall this lib, selenium, and then use playwright.
The code quality there is MUCH better, has FULL type annonation support, has very feature rich event supports (you can actually modify what browser is sending to the website, without proxying the browser and messing up with TLS Fingerprintings!)

Indeed it has sync api, but i think tbh its sync api is shit and you should be using the async api because most of actions here are "async" like intercepting a request, modifying request and etc. 
And ofc, you can integrate it easily to any custom browser you would like to implement/whatever

**Reply by xAffan** [reply]
Not asking you to fork this repo. Just create a new library for playwright instead? This repo can be used for inspiration.

**Reply by ManiMozaffar** [reply]
Don't use chrome, playwright is still undetectable in many websites, or else you can use custom written browser. up to u. but there's no need for a separate repo

**Reply by donggoing** [reply]
Could you please tell me what you mean by custom written browser?

**Comment by Connor9994** [comment]
Not to mention completely blocking community engagement on the broken (and very ram leaky) NoDriver. 

Why have everybody squeeze issues on every single one of your repos into this one @ultrafunkamsterdam?

**Reply by ultrafunkamsterdam** [reply]
Because it is an issue tracker. Not a support forum. 9/10 issues created are user issues instead of library issues. If you can do better, fork it and build your own non "leaky" package. Oh.. forgot. Never going to happen.

**Comment by ivanmilevtues** [comment]
Hey all,

I read through the discussion here. I want to mainly touch on @ManiMozaffar point, me and a friend recognized that it can be hard to get to know a new project and since it seems like this one doesn't have very clean architecture that makes it even harder.

We generated a diagram map for this repo and I would love to hear if you think this can be useful in more general sense - you can see it [here](https://github.com/CodeBoarding/GeneratedOnBoardings/blob/main/undetected-chromedriver/on_boarding.md) and I also opened a discussion around it: https://github.com/ultrafunkamsterdam/undetected-chromedriver/discussions/2206

**Reply by ultrafunkamsterdam** [reply]
No. You are just here to advertise your business. If you'd care you would submit PR's

**Reply by ivanmilevtues** [reply]
I am sorry that you think it this way. I had started submitting PR's however people felt like that is advertisement, so now I always first go in the discussions ot see if people think that what we do makes sense and liek the generation we've made for the repo. If this is the case I will be more than happy to make a PR.

We are not charging money for this, we are still trying to figure out what is useful for people as we want to build something which people like in the first place.
